bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/discovery/GreeDeviceFinder.java

```diff
@@ -0,0 +1,179 @@
+import com.google.gson.GsonBuilder;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * The GreeDeviceFinder provides functionality for searching for GREE Airconditioners on the network and keeping a list
+ * of found devices.
+ *
+ * @author John Cunha - Initial contribution
+ * @author Markus Michels - Refactoring, adapted to OH 2.5x
+ */
+@NonNullByDefault
+public class GreeDeviceFinder {
+    private final Logger logger = LoggerFactory.getLogger(GreeDeviceFinder.class);
+    private static final Gson gson = (new GsonBuilder()).create();
+
+    protected final InetAddress ipAddress;
+    protected Map<String, GreeAirDevice> deviceTable = new HashMap<>();
+
+    public GreeDeviceFinder() {
+        ipAddress = InetAddress.getLoopbackAddress(); // dummy
+    }
+
+    public GreeDeviceFinder(String broadcastAddress) throws GreeException {
+        try {
+            ipAddress = InetAddress.getByName(broadcastAddress);
+        } catch (UnknownHostException e) {
+            throw new GreeException("Unknown host or invalid IP address", e);
+        }
+    }
+
+    public void scan(DatagramSocket clientSocket, boolean scanNetwork) throws GreeException {
+        try {
+            byte[] sendData = new byte[1024];
+            byte[] receiveData = new byte[1024];
+
+            // Send the Scan message
+            GreeScanRequestDTO scanGson = new GreeScanRequestDTO();
+            scanGson.t = GREE_CMDT_SCAN;
+            String scanReq = gson.toJson(scanGson);
+            sendData = scanReq.getBytes(StandardCharsets.UTF_8);
+            logger.trace("Sending scan packet to {}", ipAddress.getHostAddress());
+            clientSocket.setSoTimeout(DISCOVERY_TIMEOUT_MS);
+            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, ipAddress, DISCOVERY_TIMEOUT_MS);
+            clientSocket.send(sendPacket);
+
+            // Loop for respnses from devices until we get a timeout.
+            int retries = MAX_SCAN_CYCLES;
+            while ((retries > 0)) {
+                // Receive a response
+                DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
+                try {
+                    clientSocket.receive(receivePacket);
+                    InetAddress remoteAddress = receivePacket.getAddress();
+                    int remotePort = receivePacket.getPort();
+
+                    // Read the response
+                    // String modifiedSentence = new String(receivePacket.getData(), StandardCharsets.UTF_8);
+                    // GreeScanResponseDTO scanResponseGson = gson.fromJson(modifiedSentence,
+                    // GreeScanResponseDTO.class);
+                    String modifiedSentence = new String(receivePacket.getData(), StandardCharsets.UTF_8);
+                    StringReader stringReader = new StringReader(modifiedSentence);
+                    GreeScanResponseDTO scanResponseGson = gson.fromJson(new JsonReader(stringReader),
```

Comment:
Gson already has the method that allows parsing from a string.
```suggestion
                    String modifiedSentence = new String(receivePacket.getData(), StandardCharsets.UTF_8);
                    GreeScanResponseDTO scanResponseGson = gson.fromJson(modifiedSentence,
```

Reply:
refactored